### DISCUSSION TITLE
How to turn off default mappings

### DISCUSSION THREAD

**Discussion by benbrastmckie** [kylechui/nvim-surround#173] [Q&A]
I'm also switching from the old surround plugins to this one, and would like to turn off most of the default mappings which I don't use, keeping just the few that I do. I imagine there must be a way to do this, but couldn't find anything in the existing discussion or issues.

I would also like to use the few surround functions that I use via WhichKey. Right now it works OK mapping `<space>ss` to `ys` and etc., however, if it was possible to put whatever it is that `ys` triggers in place of `ys`, that would be preferable. I see that `ys` corresponds to `normal` under `keymaps` which is under `M.default_opts` in the config, however, given my limited understanding of how lua works, I can't see my way all the way to an answer to my question. Any help would be greatly appreciated.

**Comment by kylechui** [comment]
As per `:h nvim-surround.configuration`,
```
However, to disable any default behavior, the corresponding value needs to explicitly be set to `false`.
```

In your case, set the corresponding keymaps to `false` in the setup function, i.e.
```lua
require('nvim-surround').setup({
    keymaps = {
        insert = false,
    }
})
```
The default configuration can be found in the `config.lua` file, or (hopefully) the help menu.


I don't personally use which-key, so if you could provide a bit more context on your issue, I'd be glad to help!

**Comment by benbrastmckie** [comment]
That's great, thanks so much! WhichKey is easy to configure, though I need to know what command to put in place of COMMAND below:

```
  s = {
    name = "SURROUND",
    s = { "COMMAND", "surround" },
    d = { "COMMAND", "delete" },
    c = { "COMMAND", "change" },
  },
``` 

I've tried putting `ys` etc., but want to remove these keybindings as well so that all surround features are accessible through WhichKey by hitting `<space>sx` where `x` is replaced by either `s`, `d`, or `c` as above.

Typically the commands in WhichKey look something like: `<cmd>lua _HTOP_TOGGLE()<cr>` or `<cmd>Telescope git_branches<cr>`, etc.

Thanks again for your help!

**Reply by kylechui** [reply]
Edit: For some reason these don't seem to be working at the moment; I'll look into it.

Ah I see now. I haven't actually set up any commands, but FWIW the keymaps call the following functions:
* `<Cmd>lua require("nvim-surround").normal_surround()<CR>`
* `<Cmd>lua require("nvim-surround").change_surround()<CR>`
* `<Cmd>lua require("nvim-surround").delete_surround()<CR>`

**Reply by benbrastmckie** [reply]
This is just what I need to configure with WhichKey once the functions are working. Thanks for taking a look !

**Reply by kylechui** [reply]
It seems like the way to go about this is to add `<Plug>` mappings to the plugin. I'll take some time to make sure that I'm doing this correctly, and then see if it can be merged in.

**Reply by kylechui** [reply]
I've added `<Plug>` mappings in the most recent commit; they can be found in `config.lua` but from my basic testing with `which-key` it seems that it automatically picked them up? Let me know if there are any extra questions you might have!

**Reply by benbrastmckie** [reply]
Excellent, it works great! I like how you have `nvim-surround-delete` is set up to ask what to delete just like with `nvim-surround-change`. Is there an easy way to also get the functionality where the next surround-pairs are deleted? This was the functionality I remember from the old vimscript plugin, and nice in some cases. Would be awesome to have both.

Thanks again for this great plugin!

**Reply by kylechui** [reply]
I'm afraid I'm not quite sure what you mean by this. Are you referring to something similar to what's found in `:h nvim-surround.jumps`?

**Reply by benbrastmckie** [reply]
Sorry that wasn't clear. What I had in mind is where you are located at the 'X' in the following:
```
$ \varphi \wedge [ X \psi \vee ( \chi \right arrow \gamma ) ] $
$ \varphi \wedge [ \psi \vee ( \chi X \right arrow \gamma ) ] $
```
Then if you trigger the surround-kill command that I am imagining, it would delete the most immediate containing surround-object. So in the first line, the square brackets would go away, and in the second line the round brackets would go away.

Hope that makes more sense.

**Reply by kylechui** [reply]
Ah, as of right now the `s` alias is used to target *any* built-in surrounding pair, so something like `dss` is possible.

**Reply by benbrastmckie** [reply]
I see. Is there a corresponding command that I could bind to in WhichKey? Right now I have all the aliases set to false. 

My aim here is to keep most of the additions to the vanilla vim bindings in WhichKey for the sake of new users (academics using LaTeX with no knowledge of Vim, etc.). That way they can learn the standard Vim bindings with one of those online trainer programs, and see what else they can do by navigating WhichKey. 

I may add a branch which reduces reliance on WhichKey at some point, but for now I want to get this up and running. If there is no command corresponding to `dss` that's totally fine. The command for delete works great.

**Reply by benbrastmckie** [reply]
I've started using my new config, and have noticed while working that your surround plugin is quite a bit faster than what I am used to from my vimscript config. Makes such a difference :)

**Reply by kylechui** [reply]
The aim for nvim-surround is to be extensible, so unfortunately the only keymaps are the ones provided in the setup function. Otherwise I would need to set up different keymaps for `ds}`, `ds'`, etc.

The speed thing is very interesting, considering I haven't optimized at all and know of a few places that could be better haha.

**Comment by benbrastmckie** [comment]
Ok so I created a WhichKey binding with the VimTex command `<Plug>(vimtex-delim-delete)` taken from [here](https://github.com/lervag/vimtex/blob/7a887028e91b12268fe812028621152dba549b3c/autoload/vimtex.vim#L150). It works great, deleting whatever the closest surround objects are, except for `$ ... $`. Is there a way to have one command that will delete whatever the closest surround objects are including `$`? I realise that I could specify `$` in telling it what to delete, but it would be nice if I didn't have to. Any thoughts would be greatly appreciated.

**Reply by kylechui** [reply]
Woops, completely missed this thread. You can configure your own "delete any" surround by setting an alias (should be `:h nvim-surround.config.aliases`)
